lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -3223,15 +3259,44 @@ private long prepareCommitInternal() throws IOException {
               // sneak into the commit point:
               toCommit = segmentInfos.clone();
 
+              if (anyChanges) {
+                mergeAwaitLatchRef = new AtomicReference<>();
+                MergePolicy mergeOnCommitPolicy = waitForMergeOnCommitPolicy(config.getMergePolicy(), toCommit, mergeAwaitLatchRef);
+
+                // Find any merges that can execute on commit (per MergePolicy).
+                commitMerges = mergeOnCommitPolicy.findCommitMerges(segmentInfos, this);
+                if (commitMerges != null && commitMerges.merges.size() > 0) {
+                  int mergeCount = 0;
+                  for (MergePolicy.OneMerge oneMerge : commitMerges.merges) {
+                    if (registerMerge(oneMerge)) {
+                      mergeCount++;
+                    } else {
+                      throw new IllegalStateException("MergePolicy " + config.getMergePolicy().getClass() +
```

Comment:
I If we fail to registerMerge does it necessarily mean the segment was already being merged?

Reply:
Yes -- or it means that the `OneMerge` contains a segment that doesn't exist (anymore?).

Since we execute this while holding `IndexWriter` lock, it should not be the case that the merge policy could have returned a `OneMerge` that contains a segment that no longer exists (because updating the live `SegmentInfos` should only occur within the IW lock).

Reply:
I see, thanks for explaining!